### PR TITLE
RUMM-2690 PassthroughContextMock sharing with Session Replay package

### PR DESCRIPTION
**Pull request by maciejburda**
### What and why?

Bigger version of https://github.com/DataDog/dd-sdk-ios/pull/1118
Allows sharing mocks and test utils with session replay + renames session replay to V2 naming.

### How?

This PR changes the way we use `TestUtilities` within the V1. Instead of importing we link the source to compile it. Session Replay which is first V2 package, now can depend on TestUtilities and leverage all the helpers exposed through it.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Tests/DatadogTests/Datadog/Mocks/DatadogInternal/UploadMock.swift
Comment (maciejburda) [inline Tests/DatadogTests/Datadog/Mocks/DatadogInternal/UploadMock.swift:44]: To not make all of them public, we can make the `TestUtilities` import `@testable`.

If you guys have any preference, happy to change that :)

**Comment by maxep** [comment]
Thanks @maciejburda, this is a great step in our migration effort 🙏 It all looks great, I just wonder about the folder names.
Maybe we could start aligning with the RFC and use `TestUtilities` and even `DatadogSessionReplay`, WDYT?

File: Tests/DatadogTests/Datadog/Mocks/DatadogInternal/UploadMock.swift
Comment (ncreated) [reply Tests/DatadogTests/Datadog/Mocks/DatadogInternal/UploadMock.swift:44]: I think it might make sense to do `@testable import` to limit the number of code changes in this PR. Otherwise we'll be running into git conflicts quite heavily.

File: test-utilities/Sources/TestUtilities/Mocks/CoreMocks.swift
Comment (ncreated) [inline test-utilities/Sources/TestUtilities/Mocks/CoreMocks.swift:340]: This is Crash-Reporting specific, shouldn't be in `TestUtilities` 🤔

File: test-utilities/Sources/TestUtilities/Mocks/CoreMocks.swift
Comment (ncreated) [inline test-utilities/Sources/TestUtilities/Mocks/CoreMocks.swift:240]: This is Logging-Reporting specific, shouldn't be in `TestUtilities` 🤔

**Review by ncreated — CHANGES_REQUESTED** [review]
Really well done 👍🎯.

I'd consider 2 things:
- to limit the number of "`public`" code changes in `DatadogTests`, it seems smarter to keep everything `internal` and we will leverage `@testable import TestUtilities` in Session Replay (looks easier this way)
- most of the things in `CoreMocks.swift` need to be segregated - there are some Loggging-, CrashReporting- or Core- specific things we shouldn't be moving (I'd be up for pairing with you to do this segregation 🙌)

File: TestUtilities/README.md
Comment (buraizu) [inline TestUtilities/README.md:3]:
```suggestion
Package that provides foundational tools, extensions, and mocks needed for testing other targets.
```

**Review by buraizu — CHANGES_REQUESTED** [review]

File: DatadogSessionReplay/Tests/DatadogSessionReplayTests/DatadogCoreIntegration/RequestBuilder/RequestBuilderTests.swift
Comment (ncreated) [inline DatadogSessionReplay/Tests/DatadogSessionReplayTests/DatadogCoreIntegration/RequestBuilder/RequestBuilderTests.swift:18]: nit/ We could try adding at least one actual test in SR to make sure it works end-to-end. `RequestBuilder` might not be the easiest one, but I made following try with `SR/../WriterTests` and it works perfectly well ⭐:

```diff
     func testWhenFeatureScopeIsConnected_itWritesRecordsToCore() {
-        // TODO: RUMM-2690
-        // Implementing this test requires creating mocks for `DatadogContext` (passed in `FeatureScope`),
-        // which is yet not possible as we lack separate, shared module to facilitate tests.
+        // Given
+        let core = PassthroughCoreMock()
+        let writer = Writer()
+
+        // When
+        writer.startWriting(to: core)
+
+        // Then
+        writer.write(nextRecord: EnrichedRecord(rumContext: .mockRandom(), records: .mockRandom()))
+        writer.write(nextRecord: EnrichedRecord(rumContext: .mockRandom(), records: .mockRandom()))
+        writer.write(nextRecord: EnrichedRecord(rumContext: .mockRandom(), records: .mockRandom()))
+
+        XCTAssertEqual(core.events(ofType: EnrichedRecord.self).count, 3)
     }
```

**Review by ncreated — CHANGES_REQUESTED** [review]
It looks really good and will greatly help on SR development 👍.

There are few remaining things we need to do after renaming folders:
* the SR path in linter setup needs to be fixed:
   * https://github.com/DataDog/dd-sdk-ios/blob/3fce96be274e62014687a6c83ef7dd3f15a52391/tools/lint/sources.swiftlint.yml#L92
   * https://github.com/DataDog/dd-sdk-ios/blob/3fce96be274e62014687a6c83ef7dd3f15a52391/tools/lint/tests.swiftlint.yml#L85
* we need to add `TestUtilities` path to linter (tests) configuration ☝️ 
*  need to update SR path prefix in `choose_workflows.py` CI script:
   * https://github.com/DataDog/dd-sdk-ios/blob/d19f328d27574dead3ed9439141b6f54d38c575d/tools/ci/choose_workflows.py#L43
* The `choose_workflows.py` decides which tests to run for given PR based on modified file paths. I think we should now configure it to run both main SDK unit tests and SR unit tests if there is a change in `TestUtilities` folder. Otherwise it will be easy to break one product by test changes made for another. If that happens, we will only realise it after merging to `develop` where we all tests are ran.

**Review by ncreated — CHANGES_REQUESTED** [review]

File: tools/ci/choose_workflows.py
Comment (ncreated) [inline tools/ci/choose_workflows.py:44]: If also adding this to `should_run_unit_tests` definition, we will be fully set 😉:
* if a PR makes change to SR and TU → run SDK unit tests for sanity check
* if a PR makes change to SDK and TU → run SR unit tests for sanity check

Otherwise, changing SR and TU might break SDK units and we won't notice this on PR branch.

**Review by ncreated — APPROVED** [review]
Well done 🎯👌. Left last minor feedback.

**Review by maxep — COMMENTED** [review]
Looks great, thanks @maciejburda 🚀

**Review by maxep — APPROVED** [review]

File: TestUtilities/README.md
Comment (ncreated) [reply TestUtilities/README.md]: TY! Changed ✅ - we will merge it without awaiting for re-approval

**Pull request merged**
